### PR TITLE
[SYCLomatic] Support migration of 4 cublas API

### PR DESCRIPTION
**Pull request by zhiweij1**
cublasSgelsBatched
cublasDgelsBatched
cublasCgelsBatched
cublasZgelsBatched

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 1]
```diff
@@ -2244,6 +2244,50 @@ inline void trmm(descriptor_ptr desc_ptr, oneapi::mkl::side left_right,
...
+/// \param [in] n The number of columns of \p a.
+/// \param [in] nrhs The number of columns of \p b.
+/// \param [in, out] a Array of pointers to matrices.
+/// \param [in] lda The leading dimension of \p a.
+/// \param [in, out] b Array of pointers to matrices.
+/// \param [in] ldb The leading dimension of \p b.
+/// \param [out] info A value stores the error information.
+/// \param [in] batch_size The size of the batch.
+template <typename T>
+inline void gels_batch_wrapper(descriptor_ptr desc_ptr,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: Please also consider dev_info_z parameter, and return type.

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 2]
```diff
@@ -168,6 +168,59 @@ inline constexpr std::size_t library_data_size[] = {
     8,                                    // real_int8_32
     8                                     // real_uint8_4
 };
+
+template <class func_t, typename... args_t>
+std::invoke_result_t<func_t, args_t...>
+catch_batch_error(int *has_execption, std::string api_names,
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]: adding an overloading like `template<class RetTy, class ...ArgsT>RetTy foo(RetTy(*)(ArgsT...) func);` may reduce the complexity when called.

File: clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp [hunk 3]
```diff
@@ -303,63 +279,40 @@ inline int potrs_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, int n,
...
+  if (info)
+    queue.memset(info, 0, group_size * sizeof(int));
+  sycl::event e = ::dpct::detail::catch_batch_error(
+      &has_execption, "oneapi::mkl::lapack::potrs_batch", queue, scratchpad,
+      nullptr, info, matrix_info->group_size_info,
+      static_cast<sycl::event (*)(
+          sycl::queue &, oneapi::mkl::uplo *, std::int64_t *, std::int64_t *,
+          Ty **, std::int64_t *, Ty **, std::int64_t *, std::int64_t,
+          std::int64_t *, Ty *, std::int64_t,
+          const std::vector<sycl::event> &)>(&oneapi::mkl::lapack::potrs_batch),
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp]:
```suggestion
          const std::vector<sycl::event> &)>(oneapi::mkl::lapack::potrs_batch),
```

File: clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp [hunk 4]
```diff
@@ -206,61 +206,37 @@ inline int potrf_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, int n,
...
+
+  if (info)
+    queue.memset(info, 0, group_size * sizeof(int));
+  sycl::event e = ::dpct::detail::catch_batch_error(
+      &has_execption, "oneapi::mkl::lapack::potrf_batch", queue, scratchpad,
+      nullptr, info, matrix_info->group_size_info,
+      static_cast<sycl::event (*)(
+          sycl::queue &, oneapi::mkl::uplo *, std::int64_t *, Ty **,
+          std::int64_t *, std::int64_t, std::int64_t *, Ty *, std::int64_t,
+          const std::vector<sycl::event> &)>(&oneapi::mkl::lapack::potrf_batch),
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp]:
```suggestion
          const std::vector<sycl::event> &)>(oneapi::mkl::lapack::potrf_batch),
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 5]
```diff
@@ -2249,6 +2249,94 @@ inline void trmm(descriptor_ptr desc_ptr, oneapi::mkl::side left_right,
...
+      &(matrix_info->lda_info), (Ty **)b, &(matrix_info->ldb_info), 1,
+      &(matrix_info->group_size_info), scratchpad, scratchpad_size,
+      std::vector<sycl::event>{});
+
+  return exec_queue.submit([&](sycl::handler &cgh) {
+    cgh.host_task([=, _e = e] {
+      ::dpct::detail::catch_batch_error(
+          nullptr, "oneapi::mkl::lapack::gels_batch", exec_queue, scratchpad,
+          info, dev_info, batch_size,
+          [](sycl::event _e) { _e.wait_and_throw(); }, _e);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
          [](sycl::event _e) { _e.wait_and_throw(); return false; }, _e);
```

File: clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp [hunk 6]
```diff
@@ -303,63 +279,40 @@ inline int potrs_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, int n,
...
+      scratchpad_size, std::vector<sycl::event>{});
+
   queue.submit([&](sycl::handler &cgh) {
-    cgh.depends_on(e);
-    cgh.host_task([=] {
+    cgh.host_task([=, _e = e] {
+      ::dpct::detail::catch_batch_error(
+          nullptr, "oneapi::mkl::lapack::potrs_batch", queue, scratchpad,
+          nullptr, info, matrix_info->group_size_info,
+          [](sycl::event _e) { _e.wait_and_throw(); }, _e);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp]:
```suggestion
          [](sycl::event _e) { _e.wait_and_throw(); return false; }, _e);
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 7]
```diff
@@ -168,6 +168,59 @@ inline constexpr std::size_t library_data_size[] = {
...
+    }
+  } catch (sycl::exception const &e) {
+    if (has_execption)
+      *has_execption = 1;
+    std::cerr << "Caught synchronous SYCL exception:" << std::endl
+              << "reason: " << e.what() << std::endl;
+    exec_queue.wait();
+    if (dev_info)
+      exec_queue.memset(dev_info, 0, batch_size * sizeof(int)).wait();
+    if constexpr (!std::is_void_v<std::invoke_result_t<func_t, args_t...>>) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp:219]:
```suggestion
    if constexpr (!std::is_void_v<std::invoke_result_t<func_t, args_t...>>) {
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 8]
```diff
@@ -168,6 +168,59 @@ inline constexpr std::size_t library_data_size[] = {
...
+      return {};
+    }
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
```

**Review by ziranzha — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 9]
```diff
@@ -168,6 +168,62 @@ inline constexpr std::size_t library_data_size[] = {
     8,                                    // real_int8_32
     8                                     // real_uint8_4
 };
+
+template <class func_t, typename... args_t>
+std::invoke_result_t<func_t, args_t...>
+catch_batch_error_from_functor(int *has_execption, std::string api_names,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
catch_batch_error_from_functor(int *has_execption, const std::string& api_names,
```

File: clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp [hunk 10]
```diff
@@ -206,61 +206,39 @@ inline int potrf_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, int n,
...
+      oneapi::mkl::lapack::potrf_batch_scratchpad_size<Ty>(
+          queue, &(matrix_info->uplo_info), &(matrix_info->n_info),
+          &(matrix_info->lda_info), 1, &(matrix_info->group_size_info));
+  Ty *scratchpad = sycl::malloc_device<Ty>(scratchpad_size, queue);
+  int has_execption = 0;
+
+  static const std::vector<sycl::event> empty_events{};
+  if (info)
+    queue.memset(info, 0, group_size * sizeof(int));
+  sycl::event e = ::dpct::detail::catch_batch_error_from_func_ptr<sycl::event>(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp]: static const std::string Name = "oneapi::mkl::lapack::potrf_batch";

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp (hunk 9 above)
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
catch_batch_error(int *has_execption, std::string api_names,
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 11]
```diff
@@ -168,6 +168,62 @@ inline constexpr std::size_t library_data_size[] = {
...
+    exec_queue.wait();
+    if (dev_info)
+      exec_queue.memset(dev_info, 0, batch_size * sizeof(int)).wait();
+  }
+  if (has_execption)
+    *has_execption = 1;
+}
+
+template <typename ret_t, typename... args_t>
+ret_t catch_batch_error_from_func_ptr(int *has_execption, std::string api_names,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
ret_t catch_batch_error_f(int *has_execption, std::string api_names,
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 12]
```diff
@@ -168,6 +168,62 @@ inline constexpr std::size_t library_data_size[] = {
     8,                                    // real_int8_32
     8                                     // real_uint8_4
 };
+
+template <class func_t, typename... args_t>
+std::invoke_result_t<func_t, args_t...>
+catch_batch_error_from_functor(int *has_execption, std::string api_names,
+                               sycl::queue exec_queue, void *scratchpad,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
                               sycl::queue exec_queue,
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 13]
```diff
@@ -168,6 +168,62 @@ inline constexpr std::size_t library_data_size[] = {
...
+        std::rethrow_exception(e);
+      } catch (oneapi::mkl::lapack::exception &e) {
+        std::cerr << "Exception in problem: " << ids[i] << std::endl
+                  << "reason: " << e.what() << std::endl
+                  << "info: " << e.info() << std::endl;
+        info_vec[ids[i]] = e.info();
+        i++;
+      }
+    }
+    exec_queue.wait();
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 14]
```diff
@@ -168,6 +168,62 @@ inline constexpr std::size_t library_data_size[] = {
...
+    if (dev_info)
+      exec_queue.memcpy(dev_info, info_vec.data(), batch_size * sizeof(int))
+          .wait();
+  } catch (sycl::exception const &e) {
+    std::cerr << "Caught synchronous SYCL exception:" << std::endl
+              << "reason: " << e.what() << std::endl;
+    exec_queue.wait();
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
```

File: clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp [hunk 15]
```diff
@@ -168,6 +168,62 @@ inline constexpr std::size_t library_data_size[] = {
...
+
+template <class func_t, typename... args_t>
+std::invoke_result_t<func_t, args_t...>
+catch_batch_error_from_functor(int *has_execption, std::string api_names,
+                               sycl::queue exec_queue, void *scratchpad,
+                               int *info, int *dev_info, int batch_size,
+                               func_t &&f, args_t &&...args) {
+  try {
+    if (has_execption)
+      *has_execption = 0;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lib_common_utils.hpp]:
```suggestion
```

File: clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp [hunk 16]
```diff
@@ -303,63 +281,42 @@ inline int potrs_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, int n,
...
+      oneapi::mkl::lapack::potrs_batch, queue, &(matrix_info->uplo_info),
+      &(matrix_info->n_info), &(matrix_info->nrhs_info), (Ty **)a,
+      &(matrix_info->lda_info), (Ty **)b, &(matrix_info->ldb_info),
+      (std::int64_t)1, &(matrix_info->group_size_info), (Ty *)scratchpad,
+      (std::int64_t)scratchpad_size, empty_events);
+
   queue.submit([&](sycl::handler &cgh) {
-    cgh.depends_on(e);
-    cgh.host_task([=] {
+    cgh.host_task([=, _e = e] {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/lapack_utils.hpp]:
```suggestion
    cgh.host_task([=, this, &func_name] mutable {
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 17]
```diff
@@ -2249,6 +2249,93 @@ inline void trmm(descriptor_ptr desc_ptr, oneapi::mkl::side left_right,
...
+      oneapi::mkl::lapack::gels_batch, exec_queue, &(matrix_info->trans_info),
+      &(matrix_info->m_info), &(matrix_info->n_info), &(matrix_info->nrhs_info),
+      (Ty **)a, &(matrix_info->lda_info), (Ty **)b, &(matrix_info->ldb_info),
+      (std::int64_t)1, &(matrix_info->group_size_info), (Ty *)scratchpad,
+      (std::int64_t)scratchpad_size, empty_events);
+
+  return exec_queue.submit([&](sycl::handler &cgh) {
+    cgh.host_task([=]() mutable {
+      ::dpct::detail::catch_batch_error(
+          nullptr, api_name, exec_queue, info, dev_info, batch_size,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
          nullptr, api_name, exec_queue, info, dev_info, matrix_info->group_size_info,
```

**Review by ziranzha — APPROVED** [review]

**Pull request merged**